zkevm-circuits/src/evm_circuit/execution/end_tx.rs

```diff
@@ -82,27 +93,60 @@ impl<F: Field> ExecutionGadget<F> for EndTxGadget<F> {
             None,
         );
 
+        let is_first_tx = IsEqualGadget::construct(cb, tx_id.expr(), 1.expr());
+
         // Add gas_used * effective_tip to coinbase's balance
         let coinbase = cb.query_cell();
+        cb.block_lookup(BlockContextFieldTag::Coinbase.expr(), None, coinbase.expr());
         let base_fee = cb.query_word_rlc();
-        for (tag, value) in [
-            (BlockContextFieldTag::Coinbase, coinbase.expr()),
-            (BlockContextFieldTag::BaseFee, base_fee.expr()),
-        ] {
-            cb.block_lookup(tag.expr(), None, value);
-        }
+        cb.condition(is_first_tx.expr(), |cb| {
+            cb.require_zero("base_fee is zero when tx is first tx", base_fee.expr());
+        });
+        cb.condition(1.expr() - is_first_tx.expr(), |cb| {
+            cb.block_lookup(BlockContextFieldTag::BaseFee.expr(), None, base_fee.expr());
+        });
+
         let effective_tip = cb.query_word_rlc();
-        let sub_gas_price_by_base_fee =
-            AddWordsGadget::construct(cb, [effective_tip.clone(), base_fee], tx_gas_price);
+        // anchor's gas_price is 0, so gas_price != effective_tip + base_fee
+        let sub_gas_price_by_base_fee = AddWordsGadget::construct(
+            cb,
+            [effective_tip.clone(), base_fee.clone()],
+            tx_gas_price.clone(),
+        );
         let mul_effective_tip_by_gas_used =
             MulWordByU64Gadget::construct(cb, effective_tip, gas_used.clone());
+        let mul_base_fee_by_gas_used =
+            MulWordByU64Gadget::construct(cb, base_fee.clone(), gas_used.clone());
+
+        // check gas_price == min(base_fee + gas_tip_cap, gas_fee_cap)
```

Comment:
Okay I think this is correct but it took me a while to figure that out! Because on https://eips.ethereum.org/EIPS/eip-1559: we have the following: 

```
# priority fee is capped because the base fee is filled first
priority_fee_per_gas = min(transaction.max_priority_fee_per_gas, transaction.max_fee_per_gas - block.base_fee_per_gas)
# signer pays both the priority fee and the base fee
effective_gas_price = priority_fee_per_gas + block.base_fee_per_gas
```

So this is a bit different than what is written here, but if the `base_fee` is subtracted from the formula than we do get to the same formula as in the EIP:
```
tx_gas_price = effective_tip + base_fee ->
effective_tip = tx_gas_price - base_fee

tx_gas_price == min(tx.gas_tip_cap + base_fee, tx.gas_fee_cap) -> (subtract base_fee everywhere)
tx_gas_price - base_fee == min(tx.gas_tip_cap, tx.gas_fee_cap - base_fee) ->
effective_tip = min(tx.gas_tip_cap, tx.gas_fee_cap - base_fee)
```

So checks out! But I'm wondering why you didn't simply use the equation from the EIP here? `tx_gas_price` and `effective_tip` are already linked in `sub_gas_price_by_base_fee` so the original equation should also be okay I think.

Reply:
Yeah, I found both of these two approaches are used in taiko-geth:
1. https://github.com/taikoxyz/taiko-geth/blob/73c5b3e7d77a02a1d0e4afb2b29210bbcc9ffb4f/accounts/abi/bind/backends/simulated.go#L636
2. https://github.com/taikoxyz/taiko-geth/blob/73c5b3e7d77a02a1d0e4afb2b29210bbcc9ffb4f/core/state_transition.go#L396

So, maybe it's good to align with eip